benchmark/httpapi.py

```diff
@@ -0,0 +1,215 @@
+
+    def retrieve(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return an empty result.
+        """
+        return succeed(list())
+
+    def delete(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+
+@implementer(IBackend)
+class InMemoryBackend(object):
+    """
+    The backend that simply drops all results.
+
+    :ivar dict results: Stored results by their identifiers.
+    """
+    def __init__(self):
+        self.results = {}
+
+    def store(self, result):
+        """
+        Store a single benchmarking result and return its identifier.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        id = uuid4().hex
+        self.results[id] = result
+        return succeed(id)
+
+    def retrieve(self, id):
+        """
+        Retrive a result by the given identifier.
+        """
+        try:
+            return self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return matching results.
+        """
+        matching = [
+            r for r in self.results.viewvalues()
+            if filter.viewitems() <= r.viewitems()
+        ]
+        return succeed(matching)
+
+    def delete(self, id):
+        """
+        Delete a result by the given identifier.
+        """
+        try:
+            del self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+
+class BenchmarkAPI_V1(object):
+    """
+    API for storing and accessing benchmarking results.
+
+    :ivar IBackend backend: The backend for storing the results.
+    """
+    app = Klein()
+    version = 1
+
+    def __init__(self, backend):
+        """
+        :param IBackend backend: The backend for storing the results.
+        """
+        self.backend = backend
+
+    @app.handle_errors(ResultNotFound)
+    def _not_found(self, request, failure):
+        request.setResponseCode(404)
+        return ""
+
+    @app.route("/submit", methods=['POST'])
+    def submit(self, request):
+        """
+        Store a new benchmarking result.
+
+        :param twisted.web.http.Request request: The request.
+        """
+        request.setHeader(b'content-type', b'application/json')
+        try:
+            json = loads(request.content.read())
+        except ValueError as e:
+            err(e, "failed to parse result")
+            request.setResponseCode(400)
+            return dumps({"message": e.message})
+
+        d = self.backend.store(json)
+
+        def stored(id):
+            msg("stored result with id {}".format(id))
+            result = {"version": self.version, "id": id}
+            return dumps(result)
+
+        d.addCallback(stored)
+        return d
```

Comment:
Because of my inexperience with RESTful APIs I overlooked that there were commonly accepted guidelines on designing the APIs. Now that, thank to you, I've discovered them I've updated the interface. I've implemented GET and DELETE methods and all tests are done through the HTTP API.